internal/wclayer/cim/block_cim_writer.go

```diff
@@ -0,0 +1,135 @@
+	if !cimfs.IsBlockCimSupported() {
+		return nil, fmt.Errorf("BlockCIM not supported on this build")
+	} else if layer.Type != cimfs.BlockCIMTypeSingleFile {
+		// we only support writing single file CIMs for now because in layer
+		// writing process we still need to write some files (registry hives)
+		// outside the CIM. We currently use the parent directory of the CIM (i.e
+		// the parent directory of block path in this case) for this. This can't
+		// be reliably done with the block device CIM since the block path
+		// provided will be a volume path. However, once we get rid of hive rollup
+		// step during layer import we should be able to support block device
+		// CIMs.
+		return nil, ErrBlockCIMWriterNotSupported
+	}
+
+	parentLayerPaths := make([]string, 0, len(parentLayers))
+	for _, pl := range parentLayers {
+		if pl.Type != layer.Type {
+			return nil, ErrBlockCIMParentTypeMismatch
+		}
+		parentLayerPaths = append(parentLayerPaths, filepath.Dir(pl.BlockPath))
+	}
+
+	cim, err := cimfs.CreateBlockCIM(layer.BlockPath, layer.CimName, layer.Type)
+	if err != nil {
+		return nil, fmt.Errorf("error in creating a new cim: %w", err)
+	}
+	defer func() {
+		if err != nil {
+			cErr := cim.Close()
+			if cErr != nil {
+				log.G(ctx).WithError(err).Warnf("failed to close cim after error: %s", cErr)
+			}
+		}
+	}()
+
+	// std file writer writes registry hives outside the CIM for 2 reasons.  1. We can
+	// merge the hives of this layer with the parent layer hives and then write the
+	// merged hives into the CIM.  2. When importing child layer of this layer, we
+	// have access to the merges hives of this layer.
+	sfw, err := newStdFileWriter(filepath.Dir(layer.BlockPath), parentLayerPaths)
+	if err != nil {
+		return nil, fmt.Errorf("error in creating new standard file writer: %w", err)
+	}
+
+	return &BlockCIMLayerWriter{
+		layer:        layer,
+		parentLayers: parentLayers,
+		addedFiles:   make(map[string]struct{}),
+		cimLayerWriter: &cimLayerWriter{
+			ctx:              ctx,
+			cimWriter:        cim,
+			stdFileWriter:    sfw,
+			layerPath:        filepath.Dir(layer.BlockPath),
+			parentLayerPaths: parentLayerPaths,
+		},
+	}, nil
+}
+
+// Add adds a file to the layer with given metadata.
+func (cw *BlockCIMLayerWriter) Add(name string, fileInfo *winio.FileBasicInfo, fileSize int64, securityDescriptor []byte, extendedAttributes []byte, reparseData []byte) error {
+	cw.addedFiles[name] = struct{}{}
+	return cw.cimLayerWriter.Add(name, fileInfo, fileSize, securityDescriptor, extendedAttributes, reparseData)
+}
+
+// Remove removes a file that was present in a parent layer from the layer.
+func (cw *BlockCIMLayerWriter) Remove(name string) error {
+	// set active write to nil so that we panic if layer tar is incorrectly formatted.
+	cw.activeWriter = nil
+	err := cw.cimWriter.AddTombstone(name)
```

Comment:
Is it safe to assume that block CIM is always merged and never forked?

Reply:
This particular `BlockCIMLayerWriter` never creates a forked block CIMs. It will always create a new layer CIM as a separate block CIM. So it is safe to assume that in this type's method.